hail/src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -43,11 +44,12 @@ object Compile {
       .zipWithIndex
       .foldLeft(Env.empty[IR]) { case (e, ((n, t, _), i)) => e.bind(n, In(i, t)) }))
     ir = LoweringPipeline.compileLowerer.apply(ctx, ir, optimize).asInstanceOf[IR]
+    ir = LowerArrayToStream(ir)
```

Comment:
this should just be part of the `compileLowerer`. It should be a lowering pipeline with two passes (first LowerArrayAggs, then LowerToStreams).

To this end, I think spending a bit of time writing down the invariants that need to be true after the IR is lowered will pay off greatly. I think you should write a new IRState for the output of LowerToStreams, called `EmittableStreamIRs` or something, which includes `ValueIROnly, EmittableValueIRs`, and `ArrayIRsAreStreams` (a rule which checks your stream invariants)

Reply:
Gotcha. Yeah, I already tried to add it as a pass coming after your agg lowering step in compileLowerer, but it broke in Interpret, match errors on stream nodes like MakeStream

Reply:
But that was with input state being emittableIR and output state being EmittableIR. Not sure of interplay with Interpet of these states, so maybe your suggestion addresses that issue.

Reply:
where in Interpret? Maybe FoldConstants needs to be updated to include a guard against stream nodes: `case x if x.typ.isInstanceOf[TStream] => None` 

here:

https://github.com/hail-is/hail/blob/2712baede392857993cca40f4544e5a10f6e0b4e/hail/src/main/scala/is/hail/expr/ir/FoldConstants.scala#L28-L27